### ISSUE TITLE
NotSupportedException when using static member constraints

### ISSUE THREAD

**Issue by abelbraaksma** [Microsoft/visualfsharp#1951]
Instead of either a compile-time error or valid code generation, the following example throws a `NotSupportedException`. I don't think it should. If this cannot be compiled (though I would think it can), I would expect a compile-time error.

#### Repro steps

Compile and run the following code, or copy to FSI:

    let inline (!>) (x:^a) : ^b = ((^a or ^b) : (static member op_Explicit: ^a -> ^b) x)

    let inline getMax v =
        match box v with
        | :? Int32 -> !> Int32.MaxValue
        | :? Int64 -> !> Int64.MaxValue
        | _ -> !> Byte.MaxValue   // other branches omitted for brevity

    let inline maxOrZero (x: ^a) =
        match x with
        | a when a = LanguagePrimitives.GenericOne -> 
            getMax x
        | a when -a = LanguagePrimitives.GenericOne -> 
            - (getMax x)
        | _ -> x    // other code omitted for brevity

    // throws System.NotSupportedException: Specified method is not supported.
    let testMe (x: int64) = 
        let foo = maxOrZero x
        foo
## Expected behavior

Should compile, or throw a compile-time error.

## Actual behavior

Throws a runtime error on every call.

The reflected code shows that whatever the input of the function is, there will always be such error:

```
public static long testMe(long x)
{
    if (x == 1L)
    {
        object source = x;
        if (LanguagePrimitives.IntrinsicFunctions.TypeTestGeneric<int>(source))
        {
            "Dynamic invocation of op_Explicit is not supported";
            throw new NotSupportedException();
        }
        if (!LanguagePrimitives.IntrinsicFunctions.TypeTestGeneric<long>(source))
        {
            "Dynamic invocation of op_Explicit is not supported";
            throw new NotSupportedException();
        }
        "Dynamic invocation of op_Explicit is not supported";
        throw new NotSupportedException();
    }
    else
    {
        if (-x != 1L)
        {
            return x;
        }
        object source2 = x;
        if (LanguagePrimitives.IntrinsicFunctions.TypeTestGeneric<int>(source2))
        {
            "Dynamic invocation of op_Explicit is not supported";
            throw new NotSupportedException();
        }
        if (!LanguagePrimitives.IntrinsicFunctions.TypeTestGeneric<long>(source2))
        {
            "Dynamic invocation of op_Explicit is not supported";
            throw new NotSupportedException();
        }
        "Dynamic invocation of op_Explicit is not supported";
        throw new NotSupportedException();
    }
}

```

## Known workarounds

I do not know of any workaround for the above code, unless by going dynamic, or using overloaded methods.

## Related information 

I tried this with F# 4.0, Update 3, VS2015.

**Comment by dsyme** [comment]
@abelbraaksma Thanks, yes, that needs to be fixed.

**Comment by dsyme** [comment]
Actually this is by design.  It even gives a warning on compilation:
```
a.fs(4,31): warning FS0077: Member constraints with the name 'op_Explicit' are given special status by the F# compiler as certain .NET types are implicitly augmented with this member. This may result in runtime failures if you attempt to invoke the member constraint from your own code.
```

Several member constraints are given special status like this.

**Comment by abelbraaksma** [comment]
I see, I should have heeded that warning (and make the connection to the behavior, after all the reflected code showed the comment quite clearly). Thanks for looking into it, sorry for wasting everybody's time.

**Comment by dsyme** [comment]
@abelbraaksma No problem.  It's a real wart, and we should probably give an error

**Comment by abelbraaksma** [comment]
@dsyme, hmm, yes and no: I have quite a bit of working code where `op_Explicit` gets resolved correctly, and I disabled the warning in those files (unfortunately, not possible per line/statement/class/module). I never understood the error, until now where it apparently hit me.

If at all possible, it would be nice to make it an error **if and only if** we can determine during compile time it would indeed create uncallable code (if that's the term for this).

**Comment by dsyme** [comment]
> If at all possible, it would be nice to make it an error if and only if we can determine during compile time it would indeed create uncallable code (if that's the term for this).

We should make it an error and include an operator (or LanguagePrimitive) in FSharp.Core that precisely corresponds to the most general possible use of the op_Explicit and op_Implicit operators

**Comment by abelbraaksma** [comment]
@dsyme, it seems that you have reverted your decision since? I tried the code from the original post again with VS2017 and the error is not thrown  anymore, at least not in FSI.

While the issue is closed, the resolution-by-design tag seems to have been overtaken by events and now the new design (both F# 4.0 and F# 4.1) allow this use of `op_Explicit` and it compiles correctly (the warning is still there, though, but that's probably for good reasons).

**Comment by dsyme** [comment]
@abelbraaksma ``testMe 1L`` still gives a NotSupportedException for me on master

**Comment by abelbraaksma** [comment]
@dsyme, you are right, it doesn't work. Wonder what I was testing to write what I wrote that day. So, just still "closed" and "by design" because it is a "special" treatment.

Looking over the original post again, I think I can now understand why:

* The function calls `getMax`, which does a _dynamic_ typetest and then expects a _static_ type test to succeed. 
* Since the type at the moment of the static type test is known, the compiler assumes that the static member constraint is valid
* But the compiled code cannot inline this, at least not this way, so it calls the automatically generated function that's generated as a fallback and throws `NotSupportedException` (I've always seen that appearing whenever I used static member constraints, even if it resolves properly, probably because the compiler cannot guarantee it "catches" every valid case).
